File: app/src/main/java/com/checkpoint/qr/detector/utilities/BitmapHandler.java
/*
 * Copyright 2019-2020 by Security and Safety Things GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.checkpoint.qr.detector.utilities;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Matrix;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import com.checkpoint.qr.detector.events.OnDetectionProcessEvent;
import com.checkpoint.qr.detector.events.OnInferenceCompletedEvent;
import com.checkpoint.qr.detector.rest.HelloWorldEndpoint;
import com.checkpoint.qr.detector.rest.QRDetectionEndPoint;
import com.securityandsafetythings.jumpsuite.commonhelpers.BitmapUtils;

/**
 * Class responsible for handling the messages sent to the BitmapHandlerThread.
 * <p> It handles the following message:
 * <ol>
 *   <li> {@link Message#SET_BITMAP} - In the {@link HelloWorldEndpoint}, sets the {@link Bitmap} object
 *   attached to the {@link android.os.Message}. </li>
 * </ol>
 */
public class BitmapHandler extends Handler {

    private static final String LOGTAG = BitmapHandler.class.getSimpleName();
    private Context context;

    /**
     * Constructs a BitmapHandler object.
     *
     * @param looper The {@code Looper} associated with the BitmapHandlerThread.
     */
    public BitmapHandler(final Looper looper) {
        super(looper);

    }

    /**
     * Acquires the latest bitmap and sets it in the {@link HelloWorldEndpoint}.
     *
     * @param msg The message containing a {@link Bitmap} received from the VideoPipeline.
     */
    @Override
    public void handleMessage(final android.os.Message msg) {
        final Message messageType = Message.fromOrdinal(msg.what);

        switch (messageType) {
        case SET_BITMAP:
            Bitmap currentBitmap = (Bitmap)msg.obj;

            QRDetectionEndPoint.getInstance().setImage(currentBitmap);
            new OnDetectionProcessEvent(currentBitmap).broadcastEvent();

            break;
        default:
            Log.e(LOGTAG, "Unknown message received on BitmapHandlerThread");
        }
    }


    /**
     * Enum defining the messages that the BitmapHandlerThread can process.
     */
    public enum Message {
        /**
         * Use this message to set the bitmap.
         */
        SET_BITMAP;

        private static final Message[] VALUES = values();

        /**
         * Retrieves the Message value given an ordinal value.
         *
         * @param ordinal The ordinal value representation of a {@code Message}.
         * @return A {@code Message} represented by the provided ordinal value.
         */
        private static Message fromOrdinal(final int ordinal) {
            return VALUES[ordinal];
        }
    }
}

File: app/src/main/java/com/checkpoint/qrdetector/handlers/BitmapHandler.kt
package com.checkpoint.qrdetector.handlers

import android.graphics.Bitmap
import android.os.Handler
import android.os.Looper
import android.util.Log

import com.checkpoint.qrdetector.events.OnDetectionProcessEvent
import com.checkpoint.qrdetector.utils.CacheFile

/*
* This looper thread receive a number of bitmaps peer second ,
* this number is determinate by camera capabilities
*
* */
class BitmapHandler (looper: Looper) : Handler(looper) {
    private var  counter = 0
     override fun handleMessage(msg: android.os.Message) {
        when (Message.fromOrdinal(msg.what)) {
            //Handler message
            Message.SET_BITMAP -> {
                val currentBitmap = msg.obj as Bitmap
                counter ++
                Log.e("BitmapHandler -->","RECEIVE BITMAP FROM CAMERA")
                //Evento to response bitmap for postproccesing.
                OnDetectionProcessEvent(currentBitmap).broadcastEvent()
            }
         }
    }

    enum class Message {
        //Handler message
        SET_BITMAP;
        companion object {
            private val VALUES = values()
    fun fromOrdinal(ordinal: Int): Message {
                return VALUES[ordinal]
            }
        }
    }
}